Source/DafnyRuntime/DafnyRuntime.cs

```diff
@@ -1183,28 +1202,24 @@ public partial class Helpers {
     public static void Print<G>(G g) {
       System.Console.Write(ToString(g));
     }
-    public static G Default<G>() {
-      Type ty = typeof(G);
-      // If ty is Dafny.ISequence<T> for some concrete T, use Dafny.Sequence<T> instead
-      // TODO-RS: How to generalize these mappings?
-      if (ty.IsGenericType && typeof(Dafny.ISet<>) == ty.GetGenericTypeDefinition()) {
-        ty = typeof(Dafny.Set<>).MakeGenericType(ty.GenericTypeArguments);
-      } else if (ty.IsGenericType && typeof(Dafny.ISequence<>) == ty.GetGenericTypeDefinition()) {
-        ty = typeof(Dafny.Sequence<>).MakeGenericType(ty.GenericTypeArguments);
-      } else if (ty.IsGenericType && typeof(Dafny.IMultiSet<>) == ty.GetGenericTypeDefinition()) {
-        ty = typeof(Dafny.MultiSet<>).MakeGenericType(ty.GenericTypeArguments);
-      } else if (ty.IsGenericType && typeof(Dafny.IMap<,>) == ty.GetGenericTypeDefinition()) {
-        ty = typeof(Dafny.Map<,>).MakeGenericType(ty.GenericTypeArguments);
-      }
-      System.Reflection.MethodInfo mInfo = ty.GetMethod("_DafnyDefaultValue");
-      if (mInfo != null) {
-        G g = (G)mInfo.Invoke(null, null);
-        return g;
-      } else {
-        return default(G);
-      }
+
+    public static readonly TypeDescriptor<bool> BOOL = new TypeDescriptor<bool>(false);
+    public static readonly TypeDescriptor<char> CHAR = new TypeDescriptor<char>('D');
```

Comment:
This character 'D' as the default is used elsewhere in the code -- is it important that they all be the same character? in any case, should they all refer to one declaration?

Reply:
Good point. It's used in the various compilers and in the verifier, and they must all agree. I added a named constant to `CharType`, and added comments to `DafnyRuntime.{cs,js,go}` and `TypeDescriptor.java' (the C++ compiler does not yet support type descriptors).

Btw, the reason it's not `'\0'` is that `'\0'` is a nuisance to deal with in printed output (like the test suite).